cuBlasLt Epilogue To Fuse Linear + ReLU|GeLU

mingxu1067 opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features

### PR changes
<!-- One of [ OPs | APIs | Docs | Others ] -->
OPs

### Describe
<!-- Describe what this PR does -->
1. Added `fused_gemm_epilogue_op` to compute Matmul+ ElementwiseAdd + ReLU|GeLU.
2. Added `fused_gemm_epilogue_grad_op` to compute ElementwiseAdd_grad + Matmul_grad+ [ReLU|GeLU]_grad.
3. Added a class member `fuse_gemm_epilogue` to `BuildStrategy` to `enable fuse_gemm_epilogue_pass`.
4. CUDA 11.6+ is required.
5. Usage Example
```python
class MultiFCLayer(paddle.nn.Layer):
    def __init__(self, hidden, Activation):
        super(MultiFCLayer, self).__init__()
        self.linear1 = paddle.nn.Linear(hidden, hidden)
        self.linear2 = paddle.nn.Linear(hidden, hidden)
        self.linear3 = paddle.nn.Linear(hidden, hidden)

        self.relu1 = Activation()
        self.relu2 = Activation()
        self.relu3 = Activation()

    def forward(self, x, matmul_y, ele_y):
        output = self.linear1(x)
        output = self.relu1(output)
        output = self.linear2(output)

        output1 = paddle.matmul(output, matmul_y)
        output = self.linear3(output)
        output = self.relu2(output)

        output = paddle.matmul(output, matmul_y)
        output = paddle.add(output, ele_y)
        output = self.relu3(output)
        output = paddle.add(output, output1)
        return output

paddle.enable_static()

batch = 64
seqlen = 128
hidden = 768

main_prog = paddle.static.Program()
startup_prog = paddle.static.Program()

with paddle.static.program_guard(main_prog, startup_prog):
    data = paddle.static.data(
        name="_data",
        shape=[-1, seqlen, hidden],
        dtype='float32')
    matmul_y = paddle.static.data(
        name="_matmul_y",
        shape=[1, hidden, hidden],
        dtype='float32')
    ele_y = paddle.static.data(
        name="_ele_y", shape=[hidden, ], dtype='float32')

    multi_fc_layer = MultiFCLayer(hidden, paddle.nn.ReLU)
    with paddle.static.amp.fp16_guard():
        out = multi_fc_layer(data, matmul_y, ele_y)
        loss = paddle.mean(out)
        paddle.static.append_backward(loss=loss)

build_strategy = paddle.static.BuildStrategy()
build_strategy.fuse_gemm_epilogue = True
program = paddle.static.CompiledProgram(smain_prog)
program = program.with_data_parallel(
    loss_name=loss.name,
    build_strategy=build_strategy,
    places=paddle.static.cuda_places())

# 3 subgraphs be fused to fused_gemm_epilogue_op
# 3 subgraphs be fused to fused_gemm_epilogue_grad
```

mingxu1067 committed:
Added cuBlasLtHandle_t to device context.

mingxu1067 committed:
Added fused_gemm_epilogue op.

1. Added fused_gemm_epilogue op to leverage cuBlastLt Epilogue.
2. Support fusion Act(X*Y + bias), X'dims >=2 and Y'dims shoule be 2.
2. Act currently only be supported ReLU. (Will add GeLU in the future).

mingxu1067 committed:
Added UT to fused_gemm_epilogue op.

mingxu1067 committed:
Added LinearAct Pattern

1. Added LinearAct into graph_pattern_detector.* to define (2.)'s
pattern.
2. LinearAct is used to detect act(element_add(matmul_v2(x, w), bias)).
3. act currently only support ReLU (Will support GeLU in the future).

mingxu1067 committed:
Added FuseGemmEpiloguePass

1, Added FuseGemmEpiloguePass to handle nn.Linear + Act{ReLU}
fusion (GeLU will be supported in the future).
2. Only support matmul_v2 from nn.Linear.

mingxu1067 committed:
Added pybind to BuildStrageter.fuse_gemm_epilogue_.

mingxu1067 committed:
Added UT for fuse_gemm_epilogue_pass.

mingxu1067 committed:
GeLU support and EpilogueSingleton

1. Added GeLU support to fused_gemm_epilogue op.
2. Added EpilogueSingleton to cache auxiliary pointer.
3. Added related UTs.

mingxu1067 committed:
Rename cublaslt_epilogue_opto gemm_epilogue_op.*.

mingxu1067 committed:
Added both train and infer pattern to LinearAct.

1. Added support of fwd graph with grap_ops linking to LinearAct.
2. Added related changes to fuse_gemm_epilogue_pass for above
modification.

mingxu1067 committed:
Changed CUDA requirement from 11.4 to 11.6 for fuse_gemm_epilogue_pass.

mingxu1067 committed:
Added identity activation support to gemm_epilogue_op.

mingxu1067 committed:
Added Linear Fusion (matmul_v2 + ele_add)

1. Added matmul_v2 + ele_add pattern to LinearActPattern.
2. Added matmul_v2 + ele_add support to fuse_gemm_epilogue_pass.

mingxu1067 committed:
Rename gemm_epilogue_op.* to fused_gemm_epilogue_op.*

mingxu1067 committed:
Add fused_gemm_epilogue_grad op.

1. Added fused_gemm_epilogue_grad to support backward epilogue fusion.

mingxu1067 committed:
Add UTs to fused_gemm_epilogue_grad_op.

mingxu1067 committed:
Change attribute name in fused_gemm_epilogue_grad_op for clearing.

mingxu1067 committed:
Allow DX and DBias be dispensable to fused_gemm_epilogue_grad op.

mingxu1067 committed:
Added ElementwiseAdd+Matmul+Act graph pattern detection.

mingxu1067 committed:
Fuse backward of Linear( Act(x))

1. Added backward fusion pass to Linear( Act(x)).
2. Added backward fusion pass to Linear(x).

mingxu1067 committed:
Added UTs to backward fusion of Linear(Act(x)).

mingxu1067 committed:
Complete document of arguments to fused_gemm_epilogue_op.

mingxu1067 committed:
Made arguments of some functions pass by reference.

mingxu1067 committed:
Modify code with review comments.

1. Made arguments of some function pass by reference.
2. Removed redundant code.
3. Followed Google code style to change code.

mingxu1067 committed:
Made 'const' code style be consistent

mingxu1067 committed:
Fixed random seed of python UTs.

mingxu1067 committed:
Merge branch 'develop'

1. Modified way to get cublasLt handler in device_context to be
consistent with last changes in develop.

mingxu1067 committed:
Set Compiling constrains to cuBlasLt

1. Require CUDA 11.6+
2. Remove fuse_gemm_epilogue related tests when CUDA < 11.6.

sneaxiy commented on paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc:
```diff
@@ -0,0 +1,443 @@
+
+bool FuseGemmEpiloguePass::IsGemmFromLinear_(
+    const std::vector<int64_t> &x_shape, const std::vector<int64_t> &w_shape,
+    OpDesc *matmul_v2_op) const {
```
Why it is enough for GEMM from `paddle.nn.Linear`? For example, how about `[M, 1] * [N, K]`? Is it supported by cublasLt?

sneaxiy commented on paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc:
```diff
@@ -0,0 +1,443 @@
+    OpDesc fused_gemm_epilogue_grad_op_desc(ele_add_grad_op->Op()->Block());
+    std::string activation_grad = "none";
+    fused_gemm_epilogue_grad_op_desc.SetType("fused_gemm_epilogue_grad");
+    fused_gemm_epilogue_grad_op_desc.SetInput("DOut",
```
How about `framework::GradVarName("Out")` instead of `DOut`? Same below.

mingxu1067 replied:
It seems that forward ops are not allowed to have `@` in Input/Output names.
```
addle/fluid/eager/api/generated/fluid_generated/dygraph_forward_api.h:146:87: error: stray ‘@’ in program
EagerTensor fused_gemm_epilogue_grad_dygraph_function(const egr::EagerTensor& Out@GRAD,const egr::EagerTensor& X,const egr::EagerTensor& Y, const paddle::framework::AttributeMap& attr_map)
```

sneaxiy commented on paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc:
```diff
@@ -0,0 +1,443 @@
+}
+
+ir::Graph *FuseGemmEpiloguePass::FuseLinearBwd(ir::Graph *graph,
+                                               bool is_first_gemm) const {
```
What is `is_first_gemm` for? What is the difference between the first and the others GEMM?

In the following code, I feel that `is_first_gemm == true` means the gradient of `X` is not needed?

mingxu1067 replied:
Done, change to `with_x_gradient`.

sneaxiy commented on paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc:
```diff
@@ -0,0 +1,335 @@
+    }
+
+    if (activation == "none" && auxiliary_key.size() > 0) {
+      VLOG(0)
```
How about use `PADDLE_THROW` here? This may be a bug if the code is reachable?

mingxu1067 replied:
Done

sneaxiy commented on paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc:
```diff
@@ -0,0 +1,335 @@
+  }
+};
+
+class FusedGemmEpilogueGradOpMaker : public framework::OpProtoAndCheckerMaker {
```
The op maker of the backward op is usually unnecessary.

mingxu1067 replied:
Since `FusedGemmEpilogueGrad` is an independent op, not grad op of `FusedGemmEpilogue`, the maker is necessary.
`FusedGemmEpilogueGrad` includes scope of backward to `Linear(Act(X))`, but `FusedGemmEpilogue` includes `Act(Linear(X))`.

sneaxiy commented on paddle/fluid/operators/fused/fused_gemm_epilogue_op.h:
```diff
@@ -0,0 +1,50 @@
+  memory::allocation::AllocationPtr auxiliary = nullptr;
+};
+
+class EpilogueSingleton {
```
In my understanding, `EpilogueSingleton` is used to store a memory buffer which is written in the forward cublasLt API. This memory buffer must be passed to the backward cublasLt API without any modification. Therefore, you use a map to save the name-to-memory-buffer mapping here, and the name is the activation output name. Am I right?

I prefer to using something like `ReserveSpace` in batch_norm op. It is not encouraged to save the variable name inside the op attribute, which makes the graph dependency analysis, .etc difficult.

mingxu1067 replied:
Done

mingxu1067 committed:
Code Reivew from Paddle

1. Changed arguments name is_first_gemm to without_x_gradient for
clearing.
2. Applied PADDLE_THROW in fused_gemm_epilogue_op.

mingxu1067 committed:
Remove EpilogueSingleton

1. Applied ReserveSpace to replace Epilogue for passing auxiliary
pointers between FWD and BWD.

mingxu1067 committed:
Fix a logical error and enhance UTs.

1. Added act op count checking in UTs.
2. Fix issue to fuse backward or ReLU(Linear(X)).
3. TODO: solve GELU fusion issues.

mingxu1067 committed:
Fix Linear and GeLU fusion issues.

1. Modified graph_detech_pattern to fit with both linear wiht gelu or
relu.
2. Modified data range in Uts to allow negative values.

mingxu1067 committed:
Removed fused_gemm_epilogue_op.h.

mingxu1067 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into cublaslt_epilogue

mingxu1067 committed:
Rename namespace pten to phi.

sneaxiy commented on paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc:
```diff
@@ -0,0 +1,353 @@
+  void Make() override {
+    AddInput("X", "The input tensor X of Out = Act((X * Y) + bias).");
+    AddInput("Y", "The input tensor Y of Out = Act((X * Y) + bias).");
+    AddInput("bias", "The input tensor bias of Out = Act((X * Y) + bias).");
```
`bias`->`Bias`?

mingxu1067 replied:
Done

sneaxiy commented on paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc:
```diff
@@ -0,0 +1,353 @@
+    AddInput("Y", "The input tensor Y of Out = Act((X * Y) + bias).");
+    AddInput("bias", "The input tensor bias of Out = Act((X * Y) + bias).");
+
+    AddOutput("out", "The output tensor Out of Out = Act((X * Y) + bias).");
```
`out`->`Out`?

mingxu1067 replied:
Done

sneaxiy commented on paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc:
```diff
@@ -0,0 +1,353 @@
+    AddInput("bias", "The input tensor bias of Out = Act((X * Y) + bias).");
+
+    AddOutput("out", "The output tensor Out of Out = Act((X * Y) + bias).");
+    AddOutput("reserve_space",
```
`reserve_space`->`ReserveSpace`?

mingxu1067 replied:
Done

sneaxiy commented on paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc:
```diff
@@ -0,0 +1,353 @@
+             "The input grad tensor to Out of Out = (Act(X) * Y) + bias");
+    AddInput("X", "The input tensor X of Out = (Act(X) * Y) + bias");
+    AddInput("Y", "The input tensor Y of Out = (Act(X) * Y) + bias");
+    AddInput("reserve_space",
```
`reserve_space`->`ReserveSpace`?

mingxu1067 replied:
Done

sneaxiy commented on paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc:
```diff
@@ -0,0 +1,469 @@
+      VarDesc reserve_space(patterns::PDNodeName(scope_name, "reserve_space"));
+      auto *reserve_space_node = g->CreateVarNode(&reserve_space);
+
+      EpiloguePassActivationCache::Instance().InsertFusedActivation(
```
How about make `EpiloguePassActivationCache::Instance` to be a local variable instead of a singleton? I mean you can change the declaration of `FuseGemmEpiloguePass::FuseLinearActFwd` and `FuseGemmEpiloguePass::FuseLinearActBwd` to be:

```C++
ir::Graph *FuseGemmEpiloguePass::FuseLinearActFwd(
    ir::Graph *graph, const std::unordered_set<std::string> &act_types,
    bool is_training, bool is_act_grad_x_from_act, EpiloguePassActivationCache *cache) const;

ir::Graph *FuseGemmEpiloguePass::FuseLinearActBwd(
    ir::Graph *graph, const std::unordered_set<std::string> &act_grad_types,
    bool is_act_grad_x_from_act, const EpiloguePassActivationCache &cache) const;
```

mingxu1067 replied:
Done.
Made `EpiloguePassActivationCache`  as local variable and pass to `FuseLinearActFwd` and `FuseLinearActBwd`.
Using pointer rather than reference, due to request from pre-commit hooks.

mingxu1067 committed:
Rename name of arguments in fused_gemm_epilogue_op

1. bias -> Bias.
2. out -> Out.
3. reserve_space -> ReserveSpace.

mingxu1067 committed:
Change EpiloguePassActivationCache as local variable.

1. Removed singleton in EpiloguePassActivationCache.
2. Made EpiloguePassActivationCache as an argument to each pass
functions.

mingxu1067 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into cublaslt_epilogue

sneaxiy reviewed (approved):
LGTM.
